### PR TITLE
Damienbod migrate protobuf to WebApiContrib.Core

### PR DESCRIPTION
**Pull request by damienbod**
Should have erverything. When it's ok, we also need to create a NuGet package

### CODE REVIEW

File: src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿
+using System.Collections.Generic;
+
+namespace WebApiContrib.Core.Formatter.Protobuf
+{
+    public class ProtobufFormatterOptions
+    {
+        //public IFormatterResolver FormatterResolver { get; set; } = ContractlessStandardResolver.Instance;
```
Comment (filipw) [inline src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs]: can we remove the commented out code?

File: src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+        public HashSet<string> SupportedContentTypes { get; set; } = new HashSet<string> { "application/x-protobuf", "application/x-protobuf" };
```
Comment (filipw) [inline src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs]: I think you meant `application/protobuf` in the second case

**Comment by filipw** [comment]
Looks great! thanks

**Comment by filipw** [comment]
btw did the tests pass? because at the moment the build process doesn't run all of them :D

**Comment by damienbod** [comment]
No public async Task Post_JSON_Header() for MessagePack and Protobuf. I'll fix and add them to the build process, thanks

File: src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs (hunk 1 above)
Comment (damienbod) [inline src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs]: ok

File: src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs (hunk 2 above)
Comment (damienbod) [inline src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs]: I think we should support all 3, opinion?
application/x-protobuf
application/protobuf
application/octet-stream

Comment (damienbod) [inline src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs]: And maybe application/x-google-protobuf

Comment (filipw) [inline src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs]: personally `application/octet-stream` is a bit invasive I think for a default option, but add it if it makes sense to you

Comment (damienbod) [inline src/WebApiContrib.Core.Formatter.Protobuf/ProtobufFormatterOptions.cs]: Think so as well, be another client uses this

From SO


  	flag
	
looks like protorpc switched to application/octet-stream: github.com/google/protorpc/commit/… – Marc Gravell♦ May 28 '15 at 12:01

**Comment by damienbod** [comment]
Supporting:
application/x-protobuf
application/protobuf
application/x-google-protobuf

Not:
application/octet-stream

**Comment by damienbod** [comment]
tests are runnning, adding to build server build now

**Comment by damienbod** [comment]
Tests are also running on the server now

**Comment by filipw** [comment]
thanks mate!

**Comment by filipw** [comment]
one more thing, can you run build.bat locally, pick up the nupkg from the artifacts folder and push 1.0.0 to nuget, so that we have it out there? 

if not, I can do it too - just let me know

**Comment by damienbod** [comment]
ok

**Comment by damienbod** [comment]
https://www.nuget.org/packages/WebApiContrib.Core.Formatter.Protobuf/1.0.0

**Comment by damienbod** [comment]
Added you as a onwer

**Comment by filipw** [comment]
excellent - thnks a lot!

**Comment by damienbod** [comment]
I'll do a quick blog to show how to use it in 1.1. Tried with 2.0, compiles, but I get runtimes bugs. Migration to 2.0 might not be a trival task :)

thanks

**Comment by damienbod** [comment]
I created an example for post protobuf data

https://github.com/damienbod/AspNetCoreWebApiContribProtobufSample/blob/master/ProtobufHttpClient/Program.cs#L40-L60

It's pretty ugly, but if I use the stream directly, the Content-Type Header is not set; "application/x-protobuf"

Any ideas?

**Comment by filipw** [comment]
you could use `StreamContent` and set the header on that

**Comment by damienbod** [comment]
Can't set the Header Content-Type for StreamContent

**Comment by filipw** [comment]
huh? why not

**Comment by damienbod** [comment]
StreamContent does allow setting the media header.

This does not either.
req.Content.Headers.ContentType = new MediaTypeHeaderValue("application/protobuf");

**Comment by filipw** [comment]
what do you mean by "does not work"? throws, or what or exactly happens? 
I've been doing it for years with system.net.http, it's even part of my book 😀

**Comment by damienbod** [comment]
The Content-Type is not set in the Header request

**Comment by damienbod** [comment]
@filipw The Context was not sent, needed to change this

request.Content = new ByteArrayContent(stream.ToArray());

**Pull request merged**